tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -4959,10 +4959,39 @@ static LogicalResult EliminateBroadcastInDimTranspose(
   return success();
 }
 
-void TransposeOp::getCanonicalizationPatterns(OwningRewritePatternList& results,
-                                              MLIRContext* /*context*/) {
+// Simplify Transpose: replace Transpose with Reshape if they are equivalent
+struct TransposeSimplifier : public OpRewritePattern<TransposeOp> {
+  using OpRewritePattern<TransposeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(TransposeOp op,
+                                PatternRewriter &rewriter) const override {
+    auto operand_type = op.operand().getType().dyn_cast<RankedTensorType>();
+    auto result_type = op.getResult().getType().dyn_cast<RankedTensorType>();
+    if (!operand_type || !result_type) {
+      return failure();
+    }
+    if (!operand_type.hasStaticShape() || !result_type.hasStaticShape()) {
```

Comment:
Is this check needed?  Or will `result_type.getDimSize(i) != 1` do the right thing even if some dims are dynamic?

Reply:
Indeed that should be fine.

Reply:
If some dims are dynamic, `result_type.getDimSize(i)` will throw some assert failure.

Reply:
From what I see the implementation is:
```
int64_t ShapedType::getDimSize(unsigned idx) const {
  assert(idx < getRank() && "invalid index for shaped type");
  return getShape()[idx];
}
```

(and `getShape()` is just an accessor to an `ArrayRef`)

Reply:
ping?

Reply:
I add some comments and one more testcase.